core/src/main/java/com/linecorp/armeria/common/stream/FlatMapStreamMessage.java

```diff
@@ -0,0 +1,400 @@
+            requireNonNull(downstream, "downstream");
+            requireNonNull(function, "function");
+            requireNonNull(executor, "executor");
+            requireNonNull(completionFuture, "completionFuture");
+
+            this.downstream = downstream;
+            this.function = function;
+            this.executor = executor;
+            this.maxConcurrency = maxConcurrency;
+            this.completionFuture = completionFuture;
+            this.options = options;
+
+            childSubscribers = new HashSet<>();
+            buffer = new ArrayDeque<>();
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+
+            upstream = subscription;
+            downstream.onSubscribe(this);
+        }
+
+        @Override
+        public void onNext(T item) {
+            requireNonNull(item, "item");
+
+            if (canceled) {
+                StreamMessageUtil.closeOrAbort(item);
+                return;
+            }
+
+            pendingSubscriptions++;
+            final StreamMessage<U> newStreamMessage = function.apply(item);
+            newStreamMessage.subscribe(new FlatMapSubscriber<>(this), executor, options);
+        }
+
+        @Override
+        public void onError(Throwable cause) {
+            requireNonNull(cause, "cause");
+
+            if (canceled) {
+                return;
+            }
+            canceled = true;
+
+            completionFuture.completeExceptionally(cause);
+            cancelChildSubscribersAndBuffer();
+            downstream.onError(cause);
+        }
+
+        @Override
+        public void onComplete() {
+            if (canceled) {
+                return;
+            }
+
+            if (canComplete()) {
+                downstream.onComplete();
+                completionFuture.complete(null);
+            } else {
+                completing = true;
+            }
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                onError(new IllegalArgumentException(
+                        "n: " + n + " (expected: > 0, see Reactive Streams specification rule 3.9)"));
+                cancel();
+                return;
+            }
+
+            if (canceled) {
+                return;
+            }
+
+            if (executor.inEventLoop()) {
+                handleRequest(n);
+            } else {
+                executor.execute(() -> handleRequest(n));
+            }
+        }
+
+        private void handleRequest(long n) {
+            requestedByDownstream = LongMath.saturatedAdd(requestedByDownstream, n);
+            flush();
+
+            if (!initialized) {
+                initialized = true;
+                upstream.request(maxConcurrency);
+            }
+
+            requestAllAvailable();
+        }
+
+        @Override
+        public void cancel() {
+            upstream.cancel();
```

Comment:
Question) Do we also want to call `cancel` here in case this method is called multiple times?
Also, is there no need to reschedule to the event loop similar to `request(n)`? (this is called by the subscriber which isn't guaranteed to call from the `executor`)

```suggestion
            if (canceled) {
                return;
            }
            canceled = true;
            upstream.cancel();
```